### PR TITLE
Create dap api

### PR DESCRIPTION
**Pull request by sjlarson92**
-created images and prompt api
-refactored front end to call api from back end and pass data to components properly
-fixed and updated front end tests

### CODE REVIEW

File: react/daily-art-prompt/src/store/rootReducer.test.js [hunk 1]
```diff
@@ -1,573 +1,506 @@
...
+            image: "apiCall"
+          }
+        };
+        expect(promptImagesReducer(state, action)).toEqual([
+          ...state,
+          "apiCall"
+        ]);
+      });
+    });
+  });
```
Comment (lgc13) [inline react/daily-art-prompt/src/store/rootReducer.test.js]: what happens if the image provided is null or undefined?

File: react/daily-art-prompt/src/store/rootReducer.test.js [hunk 2]
```diff
@@ -1,573 +1,506 @@
...
-                        payload: {
-                            imageId: 1,
-                            commentId: 2
-                        }
-                    }
-                    const result = promptImagesReducer(state, action)
-                    expect(result[0].comments[0]).toEqual(state[0].comments[0])
-                })
+  describe("when the action.type equals UPDATE_PROMPT_IMAGES", () => {
+    describe("when image.id equals payload.imageId and image.liked  is false", () => {
```
Comment (lgc13) [inline react/daily-art-prompt/src/store/rootReducer.test.js]: SPAAAAAAACEEESSSSSSSSSSSSSSSSSS  S

File: react/daily-art-prompt/src/store/rootReducer.js [hunk 3]
```diff
@@ -1,186 +1,149 @@
...
-            return state
-    }
-}
+export const dateReducer = (state = null, action) => {
+  switch (action.type) {
+    case TYPES.SET_INITIAL_DATE:
+      return moment().format("YYYY-MM-DD");
+    case TYPES.UPDATE_NEXT_DATE:
+      return action.payload.date;
+    case TYPES.UPDATE_PREVIOUS_DATE:
```
Comment (lgc13) [inline react/daily-art-prompt/src/store/rootReducer.js]: both of these do the same thing. So either:

1. Make a single action 

or

2. Put both cases together:

```js
case TYPES.UPDATE_NEXT_DATE:
case TYPES.UPDATE_PREVIOUS_DATE:
      return action.payload.date;
```

File: react/daily-art-prompt/src/store/rootReducer.test.js [hunk 4]
```diff
@@ -1,573 +1,506 @@
...
+      const action = {
+        type: TYPES.UPDATE_PREVIOUS_DATE,
+        payload: {
+          date: "2020-01-10"
+        }
+      };
+      expect(dateReducer(null, action)).toEqual("2020-01-10");
+    });
+  });
+  describe("when the case type defaults", () => {
```
Comment (lgc13) [inline react/daily-art-prompt/src/store/rootReducer.test.js]: change wording please?

File: react/daily-art-prompt/src/store/rootReducer.test.js [hunk 5]
```diff
@@ -1,573 +1,506 @@
...
+});
 
+describe("promptsReducer", () => {
+  describe("when action.type is SET_INITIAL_PROMPTS", () => {
+    it("should return initial prompts", () => {
+      const state = [];
+      const action = {
+        type: TYPES.SET_INITIAL_PROMPTS,
+        payload: {
+          prompts: ["array of prompts"]
```
Comment (lgc13) [inline react/daily-art-prompt/src/store/rootReducer.test.js]: this doesn't care what prompts are (and technically they are not arrays)

File: .idea/workspace.xml [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply .idea/workspace.xml]: okey dokey

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/ImageController.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+@RestController
+@RequestMapping("/images")
+public class ImageController {
+    ImageService imageService = new ImageService();
+
+    @GetMapping("/")
+    public ResponseEntity<Image> allimages() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*");
+
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/ImageController.java]: gone!

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/ImageController.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.HttpHeaders;
+
+@RestController
+@RequestMapping("/images")
+public class ImageController {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/ImageController.java]: everything else is called image so just changed the route name

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/PromptController.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.http.HttpHeaders;
+import java.time.LocalDate;
+import java.util.HashMap;
+
+@RestController
+@RequestMapping("/prompt")
+public class PromptController {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/PromptController.java]: done

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/model/Image.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+
+public class Image {
+    private int id;
+    private String src;
+    private boolean liked;
+    private List<String> comments = new ArrayList<>();
+
+    public Image() {
+    }
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/model/Image.java]: done

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/PromptController.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    PromptRepository promptRepository = new PromptRepository();
+
+    @GetMapping("/all")
+    public ResponseEntity<HashMap<LocalDate, Prompt>> getAllPrompts() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*");
+        HashMap<LocalDate, Prompt> responseBody = promptRepository.getAllPrompts();
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/controller/PromptController.java]: done

File: java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/service/PromptRepository.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package com.dap.DailyArtPrompt.service;
+
+import com.dap.DailyArtPrompt.model.Prompt;
+
+import java.util.HashMap;
+import java.time.LocalDate;
+
+public class PromptRepository {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply java/DailyArtPrompt/src/main/java/com/dap/DailyArtPrompt/service/PromptRepository.java]: taken care of :D

File: react/daily-art-prompt/src/EntryScreen.test.js [hunk 13]
```diff
@@ -1,43 +1,53 @@
...
+});
+
+describe("dispatchFunctions", () => {
+  it("should have getImages equal to getImagesAction", () => {
+    expect(dispatchFunctions.getImages).toEqual(getImagesAction);
+  });
+  it("should have getPrompts equal to getPromptsAction", () => {
+    expect(dispatchFunctions.getPrompts).toEqual(getPromptsAction);
+  });
+  it("should have getDate that dispatches and action SET_INITIAL_DATE", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/EntryScreen.test.js]: changed!

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 14]
```diff
@@ -1,130 +1,117 @@
...
       const wrapper = shallow(<PromptLayout {...defaultProps} />);
-      const result = wrapper.find({ 'data-testid': 'mainContentContainer' }).props().className;
-      expect(result).toEqual('prompt-row');
-    })
-
-    describe('<button> for previousButton', () => {
-      it('renders correct text', () => {
+      const result = wrapper
+        .find({ "data-testid": "mainContentContainer" })
+        .props().className;
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: changed

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 15]
```diff
@@ -1,130 +1,117 @@
...
+      it("renders correct text", () => {
         const wrapper = shallow(<PromptLayout {...defaultProps} />);
-        expect(wrapper.find({ 'data-testid': 'previousButton' }).prop('text')).toEqual('Previous');
-      })
-      it('should call handlePreviousButtonClick with correct params', () => {
+        expect(
+          wrapper.find({ "data-testid": "previousButton" }).prop("text")
+        ).toEqual("Previous");
+      });
+      it("should call updatePreviousDate", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: when its clicked!

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 16]
```diff
@@ -1,130 +1,117 @@
...
-    describe('<Prompt>', () => {
-      describe('when prompts.length is greater than 0', () => {
-        it('should render', () => {
+        wrapper.find({ "data-testid": "previousButton" }).simulate("click");
+        expect(defaultProps.updatePreviousDate).toHaveBeenCalledWith();
+      });
+    });
+
+    describe("<Prompt>", () => {
+      describe("when prompts.length is greater than 0", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: fixed!

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 17]
```diff
@@ -1,130 +1,117 @@
...
-        text: 'this is text for defaultProps.prompts'
+        date: "2019-11-19",
+        text: "i am a prompt"
       },
       {
         id: 2,
-        date: '12-18-19',
-        text: 'i love testing my code'
+        date: "2019-01-19",
+        text: "i love testing my code"
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: .... I can neither confirm nor deny this claim

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 18]
```diff
@@ -1,130 +1,117 @@
...
-        it('should render', () => {
+        wrapper.find({ "data-testid": "previousButton" }).simulate("click");
+        expect(defaultProps.updatePreviousDate).toHaveBeenCalledWith();
+      });
+    });
+
+    describe("<Prompt>", () => {
+      describe("when prompts.length is greater than 0", () => {
+        it("should render", () => {
           const wrapper = shallow(<PromptLayout {...defaultProps} />);
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: passing a prompt in scope now

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 19]
```diff
@@ -1,130 +1,117 @@
...
+          const wrapper = shallow(<PromptLayout {...defaultProps} />);
+          const result = wrapper
+            .find({ "data-testid": "prompt" })
+            .prop("prompt");
+          expect(result).toEqual(defaultProps.prompts[defaultProps.date]);
+        });
+      });
+
+      describe("when prompts.length is 0", () => {
+        it("should not render  <Prompt>", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: THE FINAL FRONTIER!

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 20]
```diff
@@ -1,130 +1,117 @@
...
+} from "./PromptLayout.js";
+import { shallow } from "enzyme";
+import {
+  updateNextDateAction,
+  updatePreviousDateAction
+} from "../store/dispatchFunctions";
+
+describe("<PromptLayout>", () => {
   const defaultProps = {
     prompts: [
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: ... yes 👍

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 21]
```diff
@@ -1,130 +1,117 @@
...
+            .prop("prompt");
+          expect(result).toEqual(defaultProps.prompts[defaultProps.date]);
+        });
+      });
+
+      describe("when prompts.length is 0", () => {
+        it("should not render  <Prompt>", () => {
           const props = {
             ...defaultProps,
             prompts: []
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: also yes 🥇

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 22]
```diff
@@ -1,130 +1,117 @@
...
+      it("renders correct text", () => {
         const wrapper = shallow(<PromptLayout {...defaultProps} />);
-        expect(wrapper.find({ 'data-testid': 'nextButton' }).prop('text')).toEqual('Next');
-      })
-      it('should call handleNextButtonClick with correct params', () => {
+        expect(
+          wrapper.find({ "data-testid": "nextButton" }).prop("text")
+        ).toEqual("Next");
+      });
+      it("should call updateNextDate with correct params", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: no! only when the planets all align and the moon eclipses the sun! only then will this occur!

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 23]
```diff
@@ -1,130 +1,117 @@
...
+        it("should render with correct prompt prop", () => {
+          const wrapper = shallow(<PromptLayout {...defaultProps} />);
+          const result = wrapper
+            .find({ "data-testid": "prompt" })
+            .prop("prompt");
+          expect(result).toEqual(defaultProps.prompts[defaultProps.date]);
+        });
+      });
+
+      describe("when prompts.length is 0", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: changed

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 24]
```diff
@@ -1,130 +1,117 @@
...
+        wrapper.find({ "data-testid": "nextButton" }).simulate("click");
+        expect(defaultProps.updateNextDate).toHaveBeenCalledWith();
+      });
+    });
+  });
+});
+
+describe("mapStateToProps", () => {
+  it("should map date to props", () => {
+    const result = mapStateToProps({ date: "2020-01-11" });
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: inserting funny puns all throughout my code :)

File: react/daily-art-prompt/src/components/PromptLayout.test.js [hunk 25]
```diff
@@ -1,130 +1,117 @@
...
+    expect(result).toEqual({ date: "2020-01-11" });
+  });
+  it("should map prompts to props", () => {
+    const result = mapStateToProps({ prompts: [{ text: "prompt" }] });
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/components/PromptLayout.test.js]: well now its a fun fact

File: react/daily-art-prompt/src/requests/imageApi.test.js [hunk 26]
```diff
@@ -1,63 +1,57 @@
...
+
+describe("getImagesAction", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("should call the api 3 times", () => {
+    axios.get.mockResolvedValue({
+      data: {
+        message: "dogPic"
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/requests/imageApi.test.js]: you're unnecessary 🗡

File: react/daily-art-prompt/src/requests/imageApi.test.js [hunk 27]
```diff
@@ -1,63 +1,57 @@
...
+      }
+    });
+    const secondFunc = getImagesAction();
+    secondFunc(dispatch);
+    expect(axios.get).toHaveBeenCalledTimes(3);
+  });
+  it("should call api with correct params", () => {
+    axios.get.mockResolvedValue({
+      data: {
+        message: "dogPic"
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/requests/imageApi.test.js]: in a bottle? lol but yes I fixed it

File: react/daily-art-prompt/src/requests/promptsApi.test.js [hunk 28]
```diff
@@ -1,78 +1,53 @@
...
+    prompts: "array of prompt objects"
+  }
+};
+
+describe("getPromptsAction", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("should call the api 1 time", async () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/requests/promptsApi.test.js]: DELETED!

File: react/daily-art-prompt/src/requests/promptsApi.test.js [hunk 29]
```diff
@@ -1,78 +1,53 @@
...
+    getPromptsAction()(dispatch);
+    expect(axios.get).toHaveBeenCalledWith("http://localhost:8080/prompt/all");
+  });
+});
+
+describe("when api response is resolved", () => {
+  it("should call dispatch with correct parameters", () => {
+    axios.get.mockResolvedValue(response);
+    getPromptsAction()(dispatch);
+    expect(dispatch).toHaveBeenNthCalledWith(1, {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/requests/promptsApi.test.js]: indubitably

File: react/daily-art-prompt/src/store/dispatchFunctions.test.js [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+  updatePreviousDateAction
+} from "./dispatchFunctions";
+
+const dispatch = jest.fn();
+
+describe("updateNextDateAction", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  describe("if prompts does not have tomorrows date", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/dispatchFunctions.test.js]: noted

File: react/daily-art-prompt/src/store/dispatchFunctions.test.js [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+    it("should not call dispatch", () => {
+      const getState = jest.fn().mockReturnValue({
+        prompts: {
+          "2020-01-17": "napping is my favorite hobby",
+          "2020-01-18": "i am an object"
+        },
+        date: "2020-01-18"
+      });
+      updateNextDateAction()(dispatch, getState);
+      expect(dispatch).toHaveBeenCalledTimes(0);
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/dispatchFunctions.test.js]: toHaveBeenCalled or not.toHaveBeenCalled that is the function!

File: react/daily-art-prompt/src/store/dispatchFunctions.js [hunk 32]
```diff
@@ -0,0 +1,30 @@
+import * as TYPES from "../store/actions";
+import moment from "moment";
+
+export const updateNextDateAction = () => (dispatch, getState) => {
+  const tomorrow = moment(getState().date)
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/dispatchFunctions.js]: done

File: react/daily-art-prompt/src/store/rootReducer.test.js [hunk 33]
```diff
@@ -1,573 +1,506 @@
...
-                    }
-                }
-                expect(promptImagesReducer(state, action)).toEqual(state)
-            })
-        })
-    })
+describe("promptImagesReducer", () => {
+  describe("when action.type equals SET_INITIAL_IMAGES", () => {
+    describe("when there is no images in state", () => {
+      it("should return updatedImages with correct attributes", () => {
```
[the remark this answers is not in the document]
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: you got it

File: react/daily-art-prompt/src/store/rootReducer.test.js (hunk 1 above)
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: the ground splits open and Armageddon begins!

but I added the test and then added an if statement to catch the condition if it is null/undefined

File: react/daily-art-prompt/src/store/rootReducer.test.js (hunk 2 above)
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: AAAAAAAHHHHHHHHHHHHHHHHH!!!

Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: I had them removed

File: react/daily-art-prompt/src/store/rootReducer.js (hunk 3 above)
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.js]: the issue has been taken care of

File: react/daily-art-prompt/src/store/rootReducer.test.js (hunk 4 above)
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: changed

File: react/daily-art-prompt/src/store/rootReducer.test.js (hunk 5 above)
Comment (sjlarson92) [reply react/daily-art-prompt/src/store/rootReducer.test.js]: well technically!!!! ... fine i changed it

**Pull request merged**